Apply registered metadata on the server

brianhelba opened the pull request:
Before being merged, this will require a database update on the live server, for commit https://github.com/ImageMarkup/isic-archive/pull/342/commits/88f25acefd6634837b6aae9079cb689bc7708e23

brianhelba commented:
@msmolens I've rebased this again, to incorporate new the `IsicResource._decodeParams` method.

msmolens commented on server/api/dataset.py:
```diff
@@ -248,6 +250,33 @@ def submitReviewImages(self, dataset, params):
+            {
+                'file': File.load(
+                    registration['fileId'], force=True, exc=True,
+                    fields=['_id', 'name']),
```
Specifying `fields` causes an error in the Girder version we use:
```
  File "/home/vagrant/girder/girder/utility/acl_mixin.py", line 63, in load
    loadType = doc['attachedToType']
KeyError: 'attachedToType'
```
To fix it I think we'd need https://github.com/girder/girder/commit/936b0ba9ea10264e5a6f0bae1516d804fd061cfd and perhaps the followup commit https://github.com/girder/girder/commit/8b10a0807214a14319682e9c6f9f809f7ccac538.

msmolens commented on server/models/dataset.py:
```diff
@@ -326,9 +328,124 @@ def registerMetadata(self, dataset, csvFile, user, sendMail=False):
+        assetstore = Assetstore.getCurrent()
+        assetstoreAdapter = assetstore_utilities.getAssetstoreAdapter(
+            assetstore)
+        metadataFilePath = assetstoreAdapter.fullPath(metadataFile)
```
The abandoned PR (https://github.com/ImageMarkup/isic-archive/pull/310) includes a change to support general assetstores. We could pull a version of that in at some point.

brianhelba commented:
I think we should add a button to directly download the currently selected CSV file, as a convenience.

brianhelba commented:
We should also add a notation to the registered metadata, once it's been applied (and perhaps containing the application time),

msmolens commented:
I think the Apply metadata view should also show a Dataset info box like on the Register metadata view. I'll add this.

brianhelba commented on web_external/Datasets/ApplyMetadataView.js:
```diff
@@ -0,0 +1,261 @@
+        var time = girder.formatDate(this.get('time'), girder.DATE_SECOND);
+        var fileName = this.get('file').name();
+        var userName = this.get('user').name();
+        return time + ' \u2014 ' + fileName + ' \u2014 ' + userName;
```
Why use a unicode "EM DASH", rather than a simple ASCII dash?

msmolens replied:
Em dash is much wider than a hyphen, and the role of the character here is to be a helpful separator between the components. A more advanced user interface here than a select box could be appropriate, and that would avoid using characters to separate information.

brianhelba replied:
Sounds good.

brianhelba commented on web_external/Datasets/ApplyMetadataView.js:
```diff
@@ -0,0 +1,261 @@
+});
+
+// Model for a single metadata error
+isic.models.MetadataErrorModel = girder.Model.extend({
```
Any reason why this is a `girder` model, rather than a `Backbone` one?

msmolens replied:
In this case, it's probably better as a `Backbone` model. I will update the code.

brianhelba commented on web_external/Datasets/ApplyMetadataView.js:
```diff
@@ -0,0 +1,261 @@
+
+// Collection of metadata errors. The list of items in the collection is
+// meaningful only when initialized() is true.
+isic.collections.MetadataErrorCollection = girder.Collection.extend({
```
Any reason why this is a `girder` collection, rather than a `Backbone` one?

msmolens replied:
In this case, it's probably better as a `Backbone` collection. I will update the code.

brianhelba commented:
@msmolens Ok, I've made the changes that we discussed, fixed testing, and cleaned up a few other small areas.

I still need to finish reviewing the content of `ApplyMetadataView.js`, which I'll do now. Please take a look at my changes.

The actual server-side logic for applying metadata is still totally broken, but I'm fine with merging it as-is (maybe we'll just change the access decorator to be admin-only). We'll clean up the server logic in a subsequent PR soon.

msmolens commented on plugin_tests/upload_test.py:
```diff
@@ -204,3 +211,114 @@ def testUploadDataset(self):
+                '_id': str(uploaderUser['_id']),
+                'name': User.obfuscatedName(uploaderUser)
+            },
+            # This is is actually checked above
```
Typo: duplicate "is"

brianhelba commented on web_external/Datasets/ApplyMetadataView.js:
```diff
@@ -0,0 +1,261 @@
+    },
+
+    fileChanged: function (file) {
+        this.file.set(file.attributes);
```
In general, I'd like to move away from this sort of thing, where we make copies of models, rather than more globally sharing a single instance. Of course, I see why you did this here, and it may indeed be the simplest way to handle it.

I wonder about the possibilities of either:
* Move the instantiation of `ApplyMetadataValidationView` into a separate method which gets called (effectively) within `filesChanged`. The `ApplyMetadataValidationView` would take a `file` as a parameter, then be destroyed and re-instantiated every time the `file` selection changes.
* Have `ApplyMetadataValidationView` expose a method to change its own file, re-rendering itself when that occurs.

I'd like to talk this through a bit further, just to be certain we're doing things the best way here.

msmolens replied:
I agree that there should be no need to duplicate models. I like your second solution above--adding a set method to triggers rerendering. However now that `ApplyMetadataValidationView` is a simple wrapper around a template, I thought it made sense to remove the view entirely and render that template in the top view. See https://github.com/ImageMarkup/isic-archive/pull/342/commits/bf6dd60acd83061df4354f6ea5fd24725043e846.

brianhelba commented:
@msmolens I've pushed a couple minor things, as well as a larger change in https://github.com/ImageMarkup/isic-archive/pull/342/commits/54b61d5e220b4f877134829fe83a25e4d53de601 . PTAL at that change, and if you have any objections, we can discuss and modify or revert it.

I have one additional minor question inline, but otherwise, this LGTM.

brianhelba commented on web_external/Datasets/applyMetadataPage.jade:
```diff
@@ -0,0 +1,25 @@
+        Once validation succeeds, you will be given an option to save the metadata.
+      .isic-apply-metadata-validate-button-container.text-center
+        button#isic-apply-metadata-validate-button.btn.btn-primary.btn-md(disabled) Validate
+      .isic-apply-metadata-validation-container
```
Why is it necessary to have two levels of `<div>` here? Couldn't we just put the `#isic-apply-metadata-validation-container` at the upper level, and dispense with the `.isic-apply-metadata-validation-container` entirely?

msmolens commented on web_external/Datasets/ApplyMetadataView.js:
```diff
@@ -143,6 +147,9 @@ isic.views.ApplyMetadataView = isic.View.extend({
 
+    /**
+     * @param {isic.models.DatasetModel} settings.dataset
+     */
```
If we're feeling ambitious we could consider using JSDoc format consistently throughout the client code.

msmolens commented:
@brianhelba I pushed a few minor changes, and don't have any more planned.
Some improvements that could go in later could be:

- After applying metadata, redirect back to dataset view, open to the correct dataset, instead of the front page. This would also apply to the upload dataset and register metadata views.
- Consider showing metadata file information on the dataset page.
